Maintain custom response schema

jaschaephraim opened the pull request:
If a schema has a custom prototype (like when using a Zod schema), normalization will be skipped. Currently this does not work in the case where there are only response schemas.

This PR accounts for that case and adds a test.

#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

jsumners commented on lib/schemas.js:
```diff
@@ -44,6 +44,26 @@ Schemas.prototype.getSchema = function (schemaId) {
   return this.store[schemaId]
 }
 
+function isCustomSchemaPrototype (schema) {
+  return typeof schema === 'object' && Object.getPrototypeOf(schema) !== Object.prototype
+}
+
+function hasCustomSchemaPrototype (routeSchemas) {
```
Please add jsdoc blocks to these functions.

Eomm commented on lib/schemas.js:
```diff
@@ -44,6 +44,26 @@ Schemas.prototype.getSchema = function (schemaId) {
+}
+
+function hasCustomSchemaPrototype (routeSchemas) {
+  for (const key in routeSchemas) {
```
we can't use `in` as it loops the prototype too.
We must list all the keys we are looking for

Eomm commented on lib/schemas.js:
```diff
@@ -44,6 +44,26 @@ Schemas.prototype.getSchema = function (schemaId) {
+function hasCustomSchemaPrototype (routeSchemas) {
+  for (const key in routeSchemas) {
+    if (key === 'response') {
+      for (const statusCode in routeSchemas.response) {
```
as above: no `in`

```suggestion
      for (const statusCode of Object.keys(routeSchemas.response)) {
```

Eomm commented on lib/schemas.js:
```diff
@@ -44,6 +44,26 @@ Schemas.prototype.getSchema = function (schemaId) {
+
+function hasCustomSchemaPrototype (routeSchemas) {
+  for (const key in routeSchemas) {
+    if (key === 'response') {
```
We can't check the response like this.

This won't work if the user sets a json schema as `header` and a custom `response` schema: in this case the `header` schema will not be processed because the `response` check will be true and it will bypass all the following logic.

It is necessary to run this check with another implementation

jaschaephraim replied:
Hey @Eomm thanks for the feedback. This seems to be how it currently works: if any one schema is a custom class then the whole normalization is skipped. I can address that and add additional test cases if that's not the intended behavior.

Eomm replied:
Yeah, for sure, it needs some refactoring.

We can sum up these cases:

input | output
---|---
json schema or empty checked | json schema checked
custom class checked | json schema checked
json schema or empty checked | custom class processed 🐛
custom class checked | custom class skipped 🐛

We need to fix only the two 🐛 cases without introducing other breaking 👍🏼

jaschaephraim commented:
Per the suggestion from @Eomm I've updated the PR to handle an underlying issue and added test cases.

Previously if any individual schema was an instance of a custom class then normalization would be skipped for all the route's schemas. Now they're checked and handled individually.

mcollina reviewed (approved):
lgtm
